keanu-project/src/test/java/io/improbable/keanu/algorithms/sampling/ForwardSamplerTest.java

```diff
@@ -0,0 +1,165 @@
+package io.improbable.keanu.algorithms.sampling;
+
+import io.improbable.keanu.Keanu;
+import io.improbable.keanu.KeanuRandom;
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.ConstantVertex;
+import io.improbable.keanu.vertices.LoadVertexParam;
+import io.improbable.keanu.vertices.TestGraphGenerator;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexId;
+import io.improbable.keanu.vertices.dbl.Differentiable;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.binary.DoubleBinaryOpVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.ProbabilisticDouble;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+
+@Slf4j
+public class ForwardSamplerTest {
+
+    private KeanuRandom random;
+
+    @Before
+    public void setup() {
+        random = new KeanuRandom(1);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwIfSampleFromHasUpstreamRandomVariables() {
+        GaussianVertex A = new GaussianVertex(100.0, 1);
+        DoubleVertex B = A.plus(ConstantVertex.of(5.));
+        GaussianVertex C = new GaussianVertex(B, 1.);
```

Comment:
If we can't even handle a simple model like this, does the forward sampler really have that much value?

Reply:
As discussed offline - we now allow conditional